### ISSUE TITLE
Add `FloatNew` and `FloatClosed` events

### ISSUE THREAD

**Issue by Danielkonge** [neovim/neovim#26548]
### Problem

I would like to be able to change some neovim behavior when opening a floating window via `vim.lsp.buf.hover()` (or other floating windows like Telescope or other plugin defined floating windows). 

One of my use cases for something like this is behavior like the following, where I basically only highlight the current screenline if I am moving around in the window in normal mode:
```lua
local daniel_opts = vim.api.nvim_create_augroup('DanielOpts', {  clear = true })
vim.api.nvim_create_autocmd({"InsertEnter", "FocusLost", "WinLeave", "CmdlineEnter"}, {
	group = daniel_opts,
	pattern = "*",
	callback = function()
		vim.opt.cursorlineopt = "number"
	end,
})
vim.api.nvim_create_autocmd({"InsertLeave", "FocusGained", "WinEnter", "CmdlineLeave"}, {
	group = daniel_opts,
	pattern = "*",
	callback = function()
		vim.opt.cursorlineopt = "screenline,number"
	end,
})
```
I would like to not show the screenline when I have an open floating window too. 

~I know I could check for open floats via `pumvisible` already, but~ I feel like it makes sense to have an event for this given that there already are similar events for buffers, windows, tabs etc. Also, I think quite a few popular plugins use floating windows a lot (like telescope, fzf-lua etc.), so it makes sense to make it easier to use auto commands for something most users will see a lot (floating windows).

(Note: I can see that #25550 updates `nvim_open_win`. I don't know if this makes sense to consider in that PR. I saw in the comments that there currently are some problems with auto commands, but I haven't looked into it.)

### Expected behavior

A `FloatNew` event that is similar to `WinNew` and either

-  a `FloatsClosed` (note the s) event that is a bit different than `WinClosed` in that it should only trigger when there are no more floats open. 
- or a `FloatClosed` event that is similar to `WinClosed` but triggers after the floating window is closed ~(then one can use `pumvisible` in the callback to check if any other floats are open after this float is closed)~ 

[Update: I misunderstood/misremembered what pumvisible does, so I crossed out some of the above.]

**Comment by wookayin** [comment]
> I know I could check for open floats via pumvisible already,

pumvisible() is for checking the popup-menu (for completion), not floating windows.

My two cents: One can easily check whether there is a floating window open in the current tabpage
```lua
local function has_floating_window(tabnr)
  tabnr = tabnr or 0  -- by default, the current tabpage
  for _, win in ipairs(vim.api.nvim_tabpage_list_wins(tabnr)) do
    return vim.api.nvim_win_get_config(win).relative ~= ""
  end
  return false
end
```
so adding such a condition to user autocmd events (WinNew, WinClose, etc.) won't be that difficult. And doing this would provide more flexibility to users -- for example, what if an user uses a plugin like https://github.com/dstein64/nvim-scrollview or fidget https://github.com/j-hui/fidget.nvim which are all based on floating windows, which users would usually want to exclude from "trigger when there are no more floats open." ?

**Comment by Danielkonge** [comment]
> pumvisible() is for checking the popup-menu (for completion), not floating windows.

Ah, I guess I didn't read the docs carefully for that comment. Thanks for the correction. 

> My two cents: One can easily check whether there is a floating window open in the current tabpage
> 
> ```lua
> local function has_floating_window(tabnr)
>   tabnr = tabnr or 0  -- by default, the current tabpage
>   for _, win in ipairs(vim.api.nvim_tabpage_list_wins(tabnr)) do
>     return vim.api.nvim_win_get_config(win).relative ~= ""
>   end
>   return false
> end
> ```
> so adding such a condition to user autocmd events (WinNew, WinClose, etc.) won't be that difficult. And doing this would provide more flexibility to users -- for example, what if an user uses a plugin like https://github.com/dstein64/nvim-scrollview or fidget https://github.com/j-hui/fidget.nvim which are all based on floating windows, which users would usually want to exclude from "trigger when there are no more floats open." ?

This is a pretty straightforward solution, yes. But wouldn't your solution also catch stuff like fidget? 

I guess a more precise request could be to allow `nvim_win_open` to broadcast events like listed above. Would that make sense? Though that would require some updates on the plugin side, and I am not sure if it makes sense to have the plugin author decide this (what if the user disagrees with a given plugin author's decision). 

I will have to think about this a bit more, but I still think some kind of events for floating windows could be nice to have.

**Comment by wookayin** [comment]
I edited the above code snippet `has_floating_window` because there was an error.

> But wouldn't your solution also catch stuff like fidget?

Yes, so my point is that users may want to add some "exclusion rules" in this function: e.g. `if the filetype or window's some metadata is ... then skip checking`.

While I think it might be a good idea to add FloatNew or FloatClose (a specialization of WinNew/WinClose events) even if they seem a bit redundant to WinNew/WinClose, I don't think "FloatsClosed", etc. are.

**Comment by Danielkonge** [comment]
> ```lua
> vim.api.nvim_create_autocmd('WinNew', {             
>   pattern = '*',       
>   group = augroup,     
>   callback = function()
>     local is_floating = vim.api.nvim_win_get_config(0).relative ~= ""
>     if is_floating then
>       ...
>     end
>   end,
> })
> ```
(I fixed a typo - missing `win` in `nvim_win_get_config`)

This doesn't seem to work. Presumably because we call `nvim_win_get_config(0)`, but we are never in the floating window when this auto command is called. #26430 should fix this when it is merged. But for now that is another nicety we could get by adding particular events for floating windows. (Though I guess #26430 will be ready before that anyways.)

**Comment by Danielkonge** [comment]
A small update: I can see that e.g. fidget actually calls `nvim_win_open` with `noautocmd = true`, which doesn't seem to be the case for e.g. Telescope or `vim.lsp.buf.hover()`. So if this behaves as I would expect, there is actually already a way for plugin authors to allow customization of which floating windows should call autocommands. (But as mentioned, we would want `FloatClosed` and not `FloatsClosed`.)

**Comment by glepnir** [comment]
Don't think we need this.  it is still within Win* scope. I add `WinNew` trigger in https://github.com/neovim/neovim/pull/25096 , if enter is false curwin will not the new float window so `vim.api.nvim_win_get_config(0)` not work. but i think you can use `buf` (callbakc data.buf) field as condition to check currently. if this is not enough. wait that pr add winid into callback data.

**Comment by clason** [comment]
I don't think we want this: From the point of Neovim, floating windows are just windows with special properties. It _might_ make sense to expose part of the config in the event data for `WinNew`. But there will always be all kinds of "floating windows", not just from fidget or hover and friends.

**Comment by justinmk** [comment]
> so `vim.api.nvim_win_get_config(0)` not work

Then this is a duplicate of https://github.com/neovim/neovim/issues/25844

**Comment by Danielkonge** [comment]
> Yes, exactly. We should expose as much useful data to autocmd handler-functions (and v:event) as needed, but new events aren't needed here. And in this case nvim_win_get_config() already exists so we don't even need to expose extra data, we only need a window-id.

What extra data would you want to expose here then? I can make a pull request adding some more data to `WinNew` and `WinClosed` if you had anything in mind. 

> Then this is a duplicate of #25844

I don't think this should really be considered a duplicate of that issue, but it doesn't really matter since both should be closed by #26430, so it is fine to keep this closed.
